Convert `id` tests to Catch2, update for SYCL 2020

psalz opened the pull request:
This is a comprehensive overhaul of the tests for the `sycl::id` class, using the new Catch2 testing framework idioms. The tests should now cover everything prescribed by the SYCL 2020 Revision 4 specification.

I'm trying out a new pattern here, where instead of performing a bunch of tests on a SYCL device and recording each "assertion's" result in a buffer that is then bulk-asserted on the host, I'm offloading individual expressions to the device using the `DEVICE_EVAL` macro. This, in my opinion, makes for overall much more readable test cases.

It currently only compiles for DPC++ because:
- hipSYCL does not implement all operators
- ComputeCpp does not support unnamed kernel lambdas yet (at least in the stable compiler, the experimental 2.8.0 compiler does support it); also it does not implement all operators

psalz commented:
(Looks like I forgot to add `id` to the ComputeCpp CI filter...)

psalz commented on tests/id/id.cpp:
```diff
@@ -0,0 +1,469 @@
+TEST_CASE("id provides specialized constructors for each dimensionality",
+          "[id]") {
+  using sycl::id;
+  CHECK(id<1>{5} == id<1>{5});
```
Just constructing the objects without any assertions felt weird, so I added the `CHECK`. But I knew somebody would complain about that ;-). I will go for checking the components, that's probably the best option.

keryell commented on tests/common/device_eval.h:
```diff
@@ -0,0 +1,36 @@
+    sycl_cts::util::get_cts_object::queue()                         \
+        .submit([=, &result_buf](sycl::handler& cgh) {              \
+          sycl::accessor result{result_buf, cgh, sycl::write_only}; \
+          cgh.single_task([=]() { result[0] = expr; });             \
```
```suggestion
          cgh.single_task([=] { result[0] = expr; });               \
```

keryell commented on tests/common/string_makers.h:
```diff
@@ -0,0 +1,35 @@
+};
+}  // namespace Catch
+
+#endif  // __SYCLCTS_TESTS_COMMON_STRING_MAKERS_H
```
Missing EOL

keryell commented on tests/id/id.cpp:
```diff
@@ -0,0 +1,482 @@
+  CHECK(DEVICE_EVAL(b >= a) == idh<D>::get(0, 0, 1));
+}
+
+#define COMPOUND_OP(operand_value, expr) \
```
Why a macro?

psalz replied:
Same as for `DEVICE_EVAL`, I want to pass an unevaluated `expr` into here.

psalz commented on tests/common/device_eval.h:
```diff
@@ -0,0 +1,36 @@
+ * - Operands must exist in surrounding scope ([=] capture).
+ * - No lambda expressions (requires C++20). Use DEVICE_EVAL_T instead.
+ */
+#define DEVICE_EVAL(expr) DEVICE_EVAL_T(decltype(expr), expr)
```
Not sure what you mean, how would I pass an unevaluated expression into a function/lambda?

keryell replied:
I see: no C++20. :-(

psalz commented on tests/id/id.cpp:
```diff
@@ -0,0 +1,482 @@
+    CHECK(DEVICE_EVAL(a[i]) == values[i]);
+  }
+
+#define ASSIGN_COMPONENT(operand_value, c, v) \
```
Because once you work with macros too much, everything seems like an opportunity for a macro ;-). Fixed.

psalz commented:
I've addressed some of the feedback and added a note to the common by-value semantics test case that there is some uncertainty around its definition (cf KhronosGroup/SYCL-Docs#210). Also I need to take a look at @vasilytric's `common_semantics.h` utilities and if/how they could be unified with this.

keryell commented on tests/common/device_eval.h:
```diff
@@ -0,0 +1,36 @@
+ */
+#define DEVICE_EVAL(expr) DEVICE_EVAL_T(decltype(expr), expr)
+
+#endif  // __SYCLCTS_TESTS_COMMON_DEVICE_EVAL_H
```
Interesting https://editorconfig.org/
I didn't know about it.

keryell reviewed (approved):
Good!

bader reviewed (approved):
Recent changes in the target branch conflict with this patch.
